File: frontend/src/services/docusign.ts
/*
Copyright 2024-2026 ChatterMate

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

import api from './api'

export interface DocuSignTemplate {
  templateId: string
  name: string
}

export interface AgentDocuSignConfig {
  enabled: boolean
  defaultTemplateId: string | null
}

/** Check if DocuSign is connected for the current organization */
export const checkDocuSignConnection = async () => {
  try {
    const response = await api.get('/docusign/status')
    return response.data
  } catch (error) {
    console.error('Error checking DocuSign connection:', error)
    return { connected: false }
  }
}

/** The DocuSign OAuth authorization URL (org connects via redirect) */
export const getDocuSignAuthUrl = () => {
  return `${import.meta.env.VITE_API_URL}/docusign/authorize`
}

/** Disconnect the DocuSign integration */
export const disconnectDocuSign = async () => {
  const response = await api.delete('/docusign/disconnect')
  return response.data
}

/** List the templates available in the connected DocuSign account */
export const getDocuSignTemplates = async (): Promise<DocuSignTemplate[]> => {
  try {
    const response = await api.get('/docusign/templates')
    return response.data
  } catch (error) {
    console.error('Error getting DocuSign templates:', error)
    return []
  }
}

/** Read an agent's DocuSign configuration */
export const getAgentDocuSignConfig = async (agentId: string): Promise<AgentDocuSignConfig> => {
  const response = await api.get(`/docusign/agent-config/${agentId}`)
  return response.data
}

/** Save an agent's DocuSign configuration */
export const saveAgentDocuSignConfig = async (agentId: string, config: AgentDocuSignConfig) => {
  const response = await api.post(`/docusign/agent-config/${agentId}`, config)
  return response.data
}

File: backend/app/models/docusign.py
"""
Copyright 2024-2026 ChatterMate

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
"""

from datetime import datetime

from sqlalchemy import Boolean, Column, DateTime, ForeignKey, Integer, String
from sqlalchemy.dialects.postgresql import UUID
from sqlalchemy.orm import relationship

from app.database import Base


class DocuSignToken(Base):
    """Per-organization DocuSign OAuth token. Secrets are Fernet-encrypted at
    rest (stored/read via app.services.docusign.tokens)."""
    __tablename__ = "docusign_tokens"

    id = Column(Integer, primary_key=True, index=True)
    organization_id = Column(UUID(as_uuid=True), ForeignKey("organizations.id"), nullable=False)
    access_token = Column(String, nullable=False)
    refresh_token = Column(String, nullable=False)
    token_type = Column(String, nullable=False)
    expires_at = Column(DateTime, nullable=False)
    account_id = Column(String, nullable=False)   # DocuSign account (API) id
    base_uri = Column(String, nullable=False)      # account base URI for REST calls
    created_at = Column(DateTime, default=datetime.utcnow)
    updated_at = Column(DateTime, default=datetime.utcnow, onupdate=datetime.utcnow)

    organization = relationship("Organization", back_populates="docusign_tokens")


class AgentDocuSignConfig(Base):
    """Which agent may send DocuSign envelopes, and its default template."""
    __tablename__ = "agent_docusign_configs"

    id = Column(Integer, primary_key=True, index=True)
    agent_id = Column(String, nullable=False, index=True)
    enabled = Column(Boolean, default=False, nullable=False)
    default_template_id = Column(String, nullable=True)
    created_at = Column(DateTime, default=datetime.utcnow)
    updated_at = Column(DateTime, default=datetime.utcnow, onupdate=datetime.utcnow)
